### PR TITLE
Implement 'action-selector get' command

### LINKED ISSUE #19: Implement 'action-selector get' command
Idea is to list members and groups existing in instance of the extern.

### PR DESCRIPTION
**Pull request by tatry**
Closes #19 

TODO:
- [x] Get whole extern (members, groups, empty group action).
- [x] Get single member.
- [x] Get single group (with members).
- [x] Get empty group action.
- [x] Tests.
- [x] Rename `default_group_action` into `empty_group_action` (to be conformant with PSA spec).

PTF tests: https://github.com/tatry/p4c/pull/12

Example JSON outputs (in all the examples Action Selector has the same content):
- Get whole extern:
  ```json
  {
      "ingress_as": {
          "member_refs": {
              "6": {
                  "action_id": 1,
                  "action_name": "ingress_a1",
                  "action_parameters": [
                      {
                          "name": "param",
                          "value": "0x6"
                      }
                  ]
              },
              "4": {
                  "action_id": 1,
                  "action_name": "ingress_a1",
                  "action_parameters": [
                      {
                          "name": "param",
                          "value": "0x4"
                      }
                  ]
              },
              "2": {
                  "action_id": 1,
                  "action_name": "ingress_a1",
                  "action_parameters": [
                      {
                          "name": "param",
                          "value": "0x2"
                      }
                  ]
              },
              "3": {
                  "action_id": 2,
                  "action_name": "ingress_a2",
                  "action_parameters": [
                      {
                          "name": "param2",
                          "value": "0x3"
                      },
                      {
                          "name": "param",
                          "value": "0x2c"
                      }
                  ]
              },
              "1": {
                  "action_id": 0,
                  "action_name": "_NoAction",
                  "action_parameters": []
              },
              "5": {
                  "action_id": 1,
                  "action_name": "ingress_a1",
                  "action_parameters": [
                      {
                          "name": "param",
                          "value": "0x5"
                      }
                  ]
              }
          },
          "group_refs": {
              "1": {
                  "member_refs": []
              },
              "4": {
                  "member_refs": [
                      6,
                      4,
                      5
                  ]
              },
              "2": {
                  "member_refs": [
                      1
                  ]
              },
              "3": {
                  "member_refs": [
                      2,
                      3
                  ]
              }
          },
          "empty_group_action": {
              "action_id": 2,
              "action_name": "ingress_a2",
              "action_parameters": [
                  {
                      "name": "param2",
                      "value": "0x2"
                  },
                  {
                      "name": "param",
                      "value": "0x80"
                  }
              ]
          }
      }
  }
  ```
- Get single member:
  ```json
  {
      "ingress_as": {
          "member_refs": {
              "3": {
                  "action_id": 2,
                  "action_name": "ingress_a2",
                  "action_parameters": [
                      {
                          "name": "param2",
                          "value": "0x3"
                      },
                      {
                          "name": "param",
                          "value": "0x2c"
                      }
                  ]
              }
          }
      }
  }
  ```
- Get single group:
  ```json
  {
      "ingress_as": {
          "member_refs": {
              "6": {
                  "action_id": 1,
                  "action_name": "ingress_a1",
                  "action_parameters": [
                      {
                          "name": "param",
                          "value": "0x6"
                      }
                  ]
              },
              "4": {
                  "action_id": 1,
                  "action_name": "ingress_a1",
                  "action_parameters": [
                      {
                          "name": "param",
                          "value": "0x4"
                      }
                  ]
              },
              "5": {
                  "action_id": 1,
                  "action_name": "ingress_a1",
                  "action_parameters": [
                      {
                          "name": "param",
                          "value": "0x5"
                      }
                  ]
              }
          },
          "group_refs": {
              "4": {
                  "member_refs": [
                      6,
                      4,
                      5
                  ]
              }
          }
      }
  }
  ```
- Get empty group action:
  ```json
  {
      "ingress_as": {
          "empty_group_action": {
              "action_id": 2,
              "action_name": "ingress_a2",
              "action_parameters": [
                  {
                      "name": "param2",
                      "value": "0x2"
                  },
                  {
                      "name": "param",
                      "value": "0x80"
                  }
              ]
          }
      }
  }
  ```

### CODE REVIEW

File: CLI/action_selector.c [hunk 1]
```diff
@@ -146,6 +150,281 @@ static int parse_skip_keyword(int *argc, char ***argv, const char *keyword)
...
+    }
+
+    return members_root;
+}
+
+json_t *create_json_group_entry(psabpf_action_selector_context_t *ctx, psabpf_action_selector_group_context_t *group, json_t *member_refs)
+{
+    json_t *group_root = json_object();
+    json_t *members = json_array();
+    if (group_root ==NULL || members == NULL) {
```
Comment (kmateuszssak) [inline CLI/action_selector.c:281]: MIssing space

File: lib/psabpf_action_selector.c [hunk 2]
```diff
@@ -728,3 +780,243 @@ uint32_t psabpf_action_selector_get_action_id_by_name(psabpf_action_selector_con
...
+    return ret;
+}
+
+psabpf_action_selector_member_context_t *psabpf_action_selector_get_next_group_member(psabpf_action_selector_context_t *ctx,
+                                                                                      psabpf_action_selector_group_context_t *group)
+{
+    if (ctx == NULL)
+        return NULL;
+    if (ctx->group.key_size != 4 || ctx->group.value_size != 4) {
+        fprintf(stderr, "invalid group map\n");
```
Comment (kmateuszssak) [inline lib/psabpf_action_selector.c:872]: Why this is invalid?

**Review by kmateuszssak — COMMENTED** [review]
Ok.

File: lib/psabpf_action_selector.c (hunk 2 above)
Comment (tatry) [reply lib/psabpf_action_selector.c:872]: For group map:
- key is a group reference, so it is always 32 bits wide;
- value is a reference to a member (or first entry is a number of references), so it is also 32 bits wide.

File: docs/command reference.md [hunk 3]
```diff
@@ -81,14 +81,15 @@ this situation.
...
 ```shell
-psabpf-ctl action-selector add_member pipe ID ACTION_SELECTOR_NAME action ACTION [data ACTION_PARAMS]
-psabpf-ctl action-selector delete_member pipe ID ACTION_SELECTOR_NAME MEMBER_REF
-psabpf-ctl action-selector update_member pipe ID ACTION_SELECTOR_NAME MEMBER_REF action ACTION [data ACTION_PARAMS]
-psabpf-ctl action-selector create_group pipe ID ACTION_SELECTOR_NAME
-psabpf-ctl action-selector delete_group pipe ID ACTION_SELECTOR_NAME GROUP_REF
-psabpf-ctl action-selector add_to_group pipe ID ACTION_SELECTOR_NAME MEMBER_REF to GROUP_REF
-psabpf-ctl action-selector delete_from_group pipe ID ACTION_SELECTOR_NAME MEMBER_REF from GROUP_REF
-psabpf-ctl action-selector default_group_action pipe ID ACTION_SELECTOR_NAME action ACTION [data ACTION_PARAMS]
+psabpf-ctl action-selector add-member pipe ID ACTION_SELECTOR_NAME action ACTION [data ACTION_PARAMS]
```
Comment (osinstom) [inline docs/command reference.md:84]: Is `MEMBER_REF` auto-generated by psabpf?

**Review by osinstom — APPROVED** [review]
LGTM in general. I left one comment.

**Comment by osinstom** [comment]
@tatry feel free to merge when ready

File: docs/command reference.md (hunk 3 above)
Comment (tatry) [reply docs/command reference.md:84]: Yes, `MEMBER_REF` is assigned by this command. Currently it is a number printed on `stdout`, see #20.

**Pull request merged**